File: src/main/java/nl/mvdr/adventofcode/adventofcode2020/day23/CrabCupsPart1.java
package nl.mvdr.adventofcode.adventofcode2020.day23;

import java.util.stream.Stream;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import nl.mvdr.adventofcode.LinesSolver;

/**
 * Solution to the day 23 puzzle of 2020's Advent of Code:
 * <a href="https://adventofcode.com/2020/day/23">Crab Cups</a>.
 *
 * @author Martijn van de Rijdt
 */
public class CrabCupsPart1 implements LinesSolver<String> {

    private static final Logger LOGGER = LoggerFactory.getLogger(CrabCupsPart1.class);

    private final int turns;
    
    /**
     * Constructor.
     * 
     * @param turns the number of turns to perform
     */
    CrabCupsPart1(int turns) {
        super();
        this.turns = turns;
    }
    
    /** Constructor. */
    public CrabCupsPart1() {
        this(100);
    }
    
    /**
     * {@inheritDoc}
     * 
     * @return order of cups, starting clockwise of the cup labeled 1
     */
    @Override
    public String solve(Stream<String> lines) {
        return GameState.parse(lines)
                .perform(turns)
                .getOrder();
    }

    /**
     * Main method.
     * 
     * @param args commandline arguments; these are ignored
     */
    public static void main(String[] args) {
        CrabCupsPart1 instance = new CrabCupsPart1();

        String result = instance.solve("input-day23-2020.txt");

        LOGGER.info(result);
    }
}
 

File: src/main/kotlin/nl/mvdr/adventofcode/adventofcode2020/day23/CrabCupsPart1.kt
package nl.mvdr.adventofcode.adventofcode2020.day23


import io.github.oshai.kotlinlogging.KotlinLogging
import nl.mvdr.adventofcode.FunctionSolver

private val logger = KotlinLogging.logger{}

fun solvePart1(lines: Sequence<String>, turns: Int = 100): String {
    val state = CrabCupsGameState(lines.toList())
    state.perform(turns)
    return state.cupsClockwiseFrom1()
}

fun main() {
    val result = FunctionSolver(::solvePart1).solve("input-day23-2020.txt")
    logger.info { result }
}
